### PR TITLE
chore: refactor classish parsers 

### PR DESCRIPTION
**Pull request by azjezz**

### CODE REVIEW

File: tests/0109/parser-error.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+ExpectedToken(["an identifier"], Some("c"), (7, 32)) -> Parse error: unexpected token `c`, expecting an identifier on line 7 column 32
```
Comment (azjezz) [inline tests/0109/parser-error.txt:1]: this error is weird, it should be "expecting a variable"

Comment (azjezz) [reply tests/0109/parser-error.txt:1]: oh... type parser is trying to parse this! this is a bug!

Comment (azjezz) [reply tests/0109/parser-error.txt:1]: see: https://3v4l.org/IRQnS, `php-parser-rs` thinks `&` means intersection types!

Comment (azjezz) [reply tests/0109/parser-error.txt]: fixed! :)

File: src/parser/classish_statement.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+                if self.current.kind == TokenKind::Equals {
+                    self.next();
+                    value = Some(self.expression(Precedence::Lowest)?);
+                }
+
+                // TODO: Support comma-separated property declarations.
+                //       nikic/php-parser does this with a single Property statement
+                //       that is capable of holding multiple property declarations.
+                self.semi()?;
+
```
Comment (azjezz) [inline src/parser/classish_statement.rs:166]: this is pretty easy to do actually, and already planned it out, will do it in next PR :)

**Comment by azjezz** [comment]
@ryangjchandler RFR!

File: src/parser/classish.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+            }),
+            None => Ok(Statement::UnitEnum {
+                name: name.into(),
+                implements,
+                body,
+            }),
+        }
+    }
+
+    fn at_least_one_comma_seperated<T>(
```
Comment (ryangjchandler) [inline src/parser/classish.rs:238]:
```suggestion
    fn at_least_one_comma_separated<T>(
```

File: src/parser/classish.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        &mut self,
+        func: &(dyn Fn(&mut Parser) -> ParseResult<T>),
+    ) -> ParseResult<Vec<T>> {
+        let mut result: Vec<T> = vec![];
+        loop {
+            result.push(func(self)?);
+            if self.current.kind == TokenKind::Comma {
```
Comment (ryangjchandler) [inline src/parser/classish.rs:245]: Flip these round so we test unhappy first, i.e. `self.current.kind != TokenKind::Comma`

File: src/parser/params.rs [hunk 5]
```diff
@@ -8,23 +8,54 @@ use crate::{
...
         let mut params = ParamList::new();
 
         while !self.is_eof() && self.current.kind != TokenKind::RightParen {
             let mut param_type = None;
 
             let mut flags: Vec<PropertyFlag> = Vec::new();
 
-            // FIXME: You should only be able to have a combination of 1 visibility + readonly modifier.
-            while matches!(
+            if matches!(
```
Comment (ryangjchandler) [inline src/parser/params.rs:27]: This code won't support `readonly public` will it? That's totally valid PHP, there's no language restriction on what order those flags/modifiers go in.

**Review by ryangjchandler — CHANGES_REQUESTED** [review]

File: src/parser/params.rs (hunk 5 above)
Comment (azjezz) [reply src/parser/params.rs:27]: oh, nice catch, will fix it :)

Comment (azjezz) [reply src/parser/params.rs]: fixed!

**Comment by azjezz** [comment]
fixed promoted property flags order, add error for when multiple access flags are used: https://github.com/ryangjchandler/php-parser-rs/compare/52a36941b99c5637696353bcff05af71b6747659..7bd58273cb4e6900d85011139ec135a7781b69cc

File: tests/0128/parser-error.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+ExpectedToken(["an identifier"], Some("s"), (5, 25)) -> Parse error: unexpected token `s`, expecting an identifier on line 5 column 25
```
Comment (ryangjchandler) [inline tests/0128/parser-error.txt:1]: Is there some way we could improve this error? The _real_ error is that there's no type defined for the prop/param.

**Review by ryangjchandler — CHANGES_REQUESTED** [review]

File: tests/0128/parser-error.txt (hunk 6 above)
Comment (azjezz) [reply tests/0128/parser-error.txt:1]: Yes, next PR will focus more on properties and promoted properties 🤞

**Pull request merged**